An alternative way to load peripherals information

asimgunes opened the pull request:
Hi,

This implementation contains a new ability to extend SVD Viewer to support not only SVD file format but also different formats. With this implementation the retrieving logic of the peripherals information separated from `PeripheralTreeProvider` class and extracted to a different class. In addition, a new parameter `getPeripherals` introduced, which enables `svd-viewer` to load the peripherals information from a different VSCode extension during the debug session by invoking the VSCode command defined in the `getPeripherals` launch parameter.

We believe this implementation provide a new ability to `svd-viewer` and we are kindly asking for review for this pull request. We also extend the documentation and provide a basic guideline for users to extend the `svd-viewer` by implementing their own VSCode plugins.

Kind regards.
Asim Gunes

thegecko commented:
@asimgunes this is a great idea!

Please be aware that this extension has now been upstreamed back to the original which has been separated at:

https://github.com/mcu-debug/peripheral-viewer

Both extensions still live in the marketplace, but I intend to deprecate this one once this issue is resolved:

https://github.com/mcu-debug/peripheral-viewer/pull/8

Let me know if you want to keep this PR here, re-open it on [peripheral-viewer](https://github.com/mcu-debug/peripheral-viewer) or open it on both. I'm keen to not review it twice :)

Thanks!

asimgunes commented:
Hi @thegecko,

I am preparing a new pull request for https://github.com/mcu-debug/peripheral-viewer and planning to send it soon. 

You can ignore this one. 

Kind regards.

asimgunes commented:
Hi @thegecko,

Thanks for returning me back, I think I didnot get the idea of providing an API here. Actually when I started I was trying somehow registering a loader and then try to use that loader here, however I wasn't successful then return back to command invocation as implemented here. But I am open for suggestions and changes here if you can explain a little bit more about your idea.

Some notes while this request was closed:
1. I remember I optimised/fixed some parts of the code in peripheral-viewer request which are not reflected to this pull request, if you plan to merge this, let me change that part too.
2. I think you are also moving this project to eclipse cdt, if you want I can provide the update to cdt project too. I can also provcide some sample code and after the migration we wouldnt need to make changes of the sample code too.

thegecko commented:
> ... if you can explain a little bit more about your idea.

For example, this extension returns an API you can interact with already:

https://github.com/eclipse-cdt-cloud/vscode-svd-viewer/blob/main/src/browser/extension.ts#L26

One approach could be to offer an extension point in that API for other file types and write an extension for your file type. e.g. expose in the API:

```typescript
    ...
    registerFileParser(fileExtension: string, (filePath: Uri) => Promise<PeripheralNode[]>)),
    ...
```

which can be called with:

```typescript
    ...
    <svdViewerApi>.registerFileParser('myEsvdExtension', async uri => {
        // do something with contents or uri and return node array
    });
    ...
```


> I remember I optimised/fixed some parts of the code in peripheral-viewer request which are not reflected to this pull request, if you plan to merge this, let me change that part too.

yes please

> I think you are also moving this project to eclipse cdt, if you want I can provide the update to cdt project too. I can also provcide some sample code and after the migration we wouldnt need to make changes of the sample code too.

already done (check the URL above!)

asimgunes commented:
Hi @thegecko,

Ok, let me check the api suggestion while checking the pull request again. it seems a better solution.

asimgunes commented:
Hi @thegecko,

I re-visit my previous implementation and changed the integration points to work over the extension api. Could you please re-check the implementation when you are available?

thegecko commented:
Thanks @asimgunes let me check it out!

thegecko commented on docs/extending-svd-viewer.md:
```diff
@@ -0,0 +1,77 @@
+
+### Developing your extension
+
+To provide the peripherals information to SVD Viewer on debug session time, you need register your command which is going construct the peripherals information. The command will receive `DebugSession` object as an input parameter and expects to return array of type `PeripheralOptions[]`.
```
```suggestion
To provide the peripherals information to SVD Viewer on debug session time, you need register your command which is going to construct the peripherals information. The command will receive `DebugSession` object as an input parameter and expects to return array of type `PeripheralOptions[]`.
```

asimgunes replied:
It is changed.

thegecko commented on docs/extending-svd-viewer.md:
````diff
@@ -0,0 +1,77 @@
+
+```js
+import { ExtensionContext } from 'vscode';
+import { 
````
Concrete implementations aren't needed, use `type` to ensure none of the extension is included at compile time.

```suggestion
import type { 
```

asimgunes replied:
AccessType is an enum so we are unable to define it as `import type` :(

thegecko commented on docs/extending-svd-viewer.md:
```diff
@@ -0,0 +1,77 @@
+	FieldOptions, 
+	AccessType,
+	IGetPeripheralsArguments,
+    IPeripheralsProvider
```
```suggestion
        IPeripheralsProvider
```

asimgunes replied:
Changed.

thegecko commented on README.md:
````diff
@@ -71,6 +71,20 @@ Once you have the SVD file, specify the location of it in your `launch.json` usi
+```json
+{
+    ...
+    "svdPath": "${workspaceFolder}/STM32F103.customFileExtension"
````
```suggestion
    "svdPath": "${workspaceFolder}/STM32F103.<customFileExtension>"
```

asimgunes replied:
Changed.

thegecko commented on docs/appendix-types.md:
```diff
@@ -0,0 +1,91 @@
+# Appendix: Types of `svd-viewer` 
```
Generally I think it would be better extracting these into a typescript file containing the interfaces. Then reference that from the docs and include it in the code. Otherwise you need to keep 2 sources in sync.

thegecko commented on docs/extending-svd-viewer.md:
```diff
@@ -0,0 +1,77 @@
+
+### Adding peripheral-viewer to your VSCode extension
+
+You need to install eclipse-cdt-cloud/vscode-svd-viewer to access the types information. You can use `npm` or `yarn` with the following arguments described below:
```
A better approach would be to publish a small package containing the extension API export which can be consumed, would be a bit more work though.

This is what Arm do for the cmsis extension:

https://www.npmjs.com/package/@arm-software/vscode-cmsis-csolution

thegecko commented on docs/extending-svd-viewer.md:
```diff
@@ -0,0 +1,77 @@
+
+	// Invoke registerPeripheralsProvider method in eclipse-cdt.svd-viewer extension api
+	// Register 'MyExtensionProvider' for files *.myext
+    svdViewerExtension?.exports?.registerPeripheralsProvider('myext', new MyExtensionProvider());
```
This could be more resilient (e.g. attach to an onActivation event)

thegecko commented on docs/extending-svd-viewer.md:
````diff
@@ -0,0 +1,77 @@
+}
+```
+
+You can check the type definitions (`PeripheralOptions`, `PeripheralRegisterOptions`, `ClusterOptions`, `FieldOptions`, `AccessType`) from [this document](./appendix-types.md).
````
Reference source file with extracted interfaces instead?

thegecko commented on src/browser/extension.ts:
```diff
@@ -11,12 +11,13 @@ import { Commands } from '../commands';
 import { DebugTracker } from '../debug-tracker';
 import { SvdRegistry } from '../svd-registry';
 import { SvdResolver } from '../svd-resolver';
+export * from '../export';
```
This is a little ugly. A different approach could be to not export here and instead guide users to import the types directly. e.g. Have the interfaces extracted to a file (e.g. `api.ts`) and update the docs to use:

```typescript
import type * as api from from 'vscode-svd-viewer/api';
```

To be used as `api.PeripheralOptions`...

asimgunes replied:
I made the change with a little bit different approach. I moved the type definitions to a separate ile called `api.ts` but I avoid to add the `api.ts` to the root folder so I kept exporting here but encapsulated it as `api`. 

Also we are unable to use `import type` at the client because `AccessType` is an enum and this is preventing the use of `import enum`

Final client usage will be like below:

```js
import { api } from 'svd-viewer';
```

How this sounds to you?

thegecko commented on src/desktop/extension.ts:
```diff
@@ -11,12 +11,13 @@ import { Commands } from '../commands';
 import { DebugTracker } from '../debug-tracker';
 import { SvdRegistry } from '../svd-registry';
 import { SvdResolver } from '../svd-resolver';
+export * from '../export';
```
ditto on export

thegecko commented on src/export.ts:
```diff
@@ -0,0 +1,7 @@
+export { PeripheralOptions } from './views/nodes/peripheralnode';
```
As mentioned, perhaps extract these all into a common api file?

asimgunes replied:
I moved them to `api.ts` file.

thegecko commented on src/types.ts:
```diff
@@ -0,0 +1,9 @@
+import { PeripheralOptions } from './views/nodes/peripheralnode';
```
More candidates for a common api file maybe?

asimgunes replied:
I moved them to `api.ts` file.

thegecko commented on src/views/peripheral.ts:
```diff
@@ -137,23 +137,38 @@ export class PeripheralTreeForSession extends PeripheralBaseNode {
+            const decoder = new TextDecoder();
+            const data = decoder.decode(contents);
+            const provider = this.registry.getPeripheralsProvider(svdPath);
+            if(provider) {
```
```suggestion
            if (provider) {
```

asimgunes replied:
Changed.

thegecko commented on src/views/peripheral.ts:
```diff
@@ -164,8 +179,9 @@ export class PeripheralTreeForSession extends PeripheralBaseNode {
+        this.loaded = true;
 
-        this.errMessage = '';
+        //this.errMessage = '';
```
```suggestion
```

asimgunes replied:
Changed.

thegecko reviewed (changes requested):
This works well for me, nothing seems broken!

I've added a few comments about structure and exposing the API and a few styling nits.

Could you also check your formatting and indentation for the new/modified files?

asimgunes commented:
Hi @thegecko, 

I made changed related to your suggestions. There could be slight differences related to implementation details. Could you check the latest implementation in your suitable time?

>This is what Arm do for the cmsis extension:
>
> https://www.npmjs.com/package/@arm-software/vscode-cmsis-csolution

Also, example about the "cmsis extension" is interesting and I can contribute for similar approach for this extension, but we need a new repository for this implementation. Additonally, it might be logical to move the type definitions and some utility functions might be useful for users, and then reference the package in svd-viewer to avoid duplication as you mentioned in this code review. But I also prefer merging the changes up to now and leaving that work for another pull request. Is it suitable for you?

thegecko commented on README.md:
````diff
@@ -71,6 +71,20 @@ Once you have the SVD file, specify the location of it in your `launch.json` usi
+```json
+{
+    ...
+    "svdPath": "${workspaceFolder}/STM32F103.<customFileExtension>"
````
```suggestion
    "definitionPath": "${workspaceFolder}/STM32F103.<customFileExtension>"
```

thegecko commented on README.md:
````diff
@@ -71,6 +71,20 @@ Once you have the SVD file, specify the location of it in your `launch.json` usi
 }
 ```
 
+### Extending SVD Viewer
````
```suggestion
### Extending Peripheral Inspector
```

thegecko commented on README.md:
````diff
@@ -71,6 +71,20 @@ Once you have the SVD file, specify the location of it in your `launch.json` usi
+}
+```
+
+For more details about the implementation, please check the [Extending SVD Viewer](./docs/extending-svd-viewer.md) document.
````
```suggestion
For more details about the implementation, please check the [Extending Peripheral Inspector](./docs/extending-svd-viewer.md) document.
```

thegecko commented on docs/extending-svd-viewer.md:
```diff
@@ -0,0 +1,74 @@
+# Extending SVD Viewer
```
```suggestion
# Extending Peripheral Inspector
```

Also, should we rename this file?

thegecko commented on README.md:
```diff
@@ -71,6 +71,20 @@ Once you have the SVD file, specify the location of it in your `launch.json` usi
 
+### Extending SVD Viewer
+
+It is possible to extend the SVD Viewer and register new file extension providers with your VSCode extension. This method will provide reading non-svd formats and load the peripherals information in SVD Viewer. 
```
```suggestion
It is possible to extend the Peripheral Inspector and register new file extension providers with your VSCode extension. This method will provide reading non-svd formats and load the peripherals information in Peripheral Inspector. 
```

thegecko commented on docs/extending-svd-viewer.md:
```diff
@@ -0,0 +1,74 @@
+# Extending SVD Viewer
+
+It is possible to extend the SVD Viewer and register new file extension providers with your VSCode extension. This method will provide reading non-svd formats and load the peripherals information in SVD Viewer. 
```
```suggestion
It is possible to extend the Peripheral Inspector and register new file extension providers with your VSCode extension. This method will provide reading non-svd formats and load the peripherals information in Peripheral Inspector. 
```

thegecko commented on docs/extending-svd-viewer.md:
```diff
@@ -0,0 +1,74 @@
+
+It is possible to extend the SVD Viewer and register new file extension providers with your VSCode extension. This method will provide reading non-svd formats and load the peripherals information in SVD Viewer. 
+
+## Building your VSCode Extension to extend SVD Viewer
```
```suggestion
## Building your VSCode Extension to extend Peripheral Inspector
```

thegecko commented on docs/extending-svd-viewer.md:
```diff
@@ -0,0 +1,74 @@
+
+## Building your VSCode Extension to extend SVD Viewer
+
+This is a guide about how you can register new peripheral providers to SVD Viewer in your VSCode extension. Please refer to [VSCode Extension API](https://code.visualstudio.com/api) for more information about developing VSCode extensions.
```
```suggestion
This is a guide about how you can register new peripheral providers to Peripheral Inspector in your VSCode extension. Please refer to [VSCode Extension API](https://code.visualstudio.com/api) for more information about developing VSCode extensions.
```

thegecko commented on docs/extending-svd-viewer.md:
```diff
@@ -0,0 +1,74 @@
+
+### Adding peripheral-viewer to your VSCode extension
+
+You need to install eclipse-cdt-cloud/vscode-svd-viewer to access the types information. You can use `npm` or `yarn` with the following arguments described below:
```
```suggestion
You need to install eclipse-cdt-cloud/vscode-peripheral-inspector to access the types information. You can use `npm` or `yarn` with the following arguments described below:
```

thegecko commented on docs/extending-svd-viewer.md:
````diff
@@ -0,0 +1,74 @@
+
+Using with npm:
+```bash
+npm install github:eclipse-cdt-cloud/vscode-svd-viewer
````
```suggestion
npm install github:eclipse-cdt-cloud/vscode-peripheral-inspector
```

thegecko commented on docs/extending-svd-viewer.md:
````diff
@@ -0,0 +1,74 @@
+```
+Using with yarn:
+```bash
+yarn add github:eclipse-cdt-cloud/vscode-svd-viewer
````
```suggestion
yarn add github:eclipse-cdt-cloud/vscode-peripheral-inspector
```

thegecko commented on docs/extending-svd-viewer.md:
```diff
@@ -0,0 +1,74 @@
+
+### Developing your extension
+
+To provide the peripherals information to SVD Viewer on debug session time, you need register your command which is going to construct the peripherals information. The command will receive `DebugSession` object as an input parameter and expects to return array of type `PeripheralOptions[]`.
```
```suggestion
To provide the peripherals information to Peripheral Inspector on debug session time, you need register your command which is going to construct the peripherals information. The command will receive `DebugSession` object as an input parameter and expects to return array of type `PeripheralOptions[]`.
```

thegecko commented on docs/extending-svd-viewer.md:
````diff
@@ -0,0 +1,74 @@
+
+```js
+import { ExtensionContext } from 'vscode';
+import { api } from "svd-viewer";
````
```suggestion
import { api } from "peripheral-inspector";
```

thegecko reviewed (changes requested):
I started renaming SVD Viewer and svd-viewer to their new equivalents but gave up after so many comments. Can you do this in the PR? :)

Please also rebase on main and check all still compiles and works for you

asimgunes commented:
Hi @thegecko,

I rebased and update the request as the name of the package changed. Could you please review again in your suitable time?

You may observe that I have additional updates. There is an additional `api.ts` at the root folder and types are now imported from 'peripheral-inspector/api' as you can also see in the documentation. This change comes because previously I missed the scenario that if the package installed directly from Github, there is no `dist` folder in it, which is causing import problem. For a workaround, I added `api.ts` file into the root, but still keep the actual definition file under the `src` folder. In long term, I think we need a separate package that includes API information like you previously mentioned about the ARM cmsis extension, but I like to leave this work to another (follow-up) update, and if you agree, merge this with the current state. Please let me know your thoughts, I am looking forward to hear them.

thegecko commented on src/api-types.ts:
```diff
@@ -0,0 +1,83 @@
+// Peripheral Inspector API
+export interface IPeripheralInspectorAPI {
+	registerSVDFile: (expression: RegExp | string, path: string) => void;
```
nit: please use spaces not tabs

thegecko reviewed (approved):
Some indentations are wrong, but that's a minor niggle which can be addressed separately.
I also think we can make the API better by using GitHubs own npm registry. I'll craft a follow on PR to discuss that.
I'll merge this once I've done a final test locally.

thegecko commented on src/api-types.ts:
```diff
@@ -0,0 +1,83 @@
+	getPeripherals: (data: string, options: IGetPeripheralsArguments) => Promise<PeripheralOptions[]>;
+}
+
+export declare enum AccessType {
```
This can't be a `declare` it turns it into a type, not an enum and then breaks the svd parsing.
```suggestion
export enum AccessType {
```

thegecko commented on tsconfig.json:
```diff
@@ -4,6 +4,8 @@
         "module": "commonjs",
         "strict": true,
         "sourceMap": true,
+        "outDir": "dist",
+        "declaration": true,
```
The outdir breaks the maps and debugging when using webpack. Then without it the declaration files pollute the source (to be fair, they shouldn't be created anyway as you have the types file).
```suggestion
```

thegecko reviewed (changes requested):
OK, so this fails to run with svd files. please accept the suggestions and all should be well. You could also check the indentation in the api-types file ;)

asimgunes commented:
Hi Rob, thanks for your comments, I am going to check the suggestions soon and planning to return you by tomorrow. Kind regards.

thegecko reviewed (approved):
LGTM

thegecko commented:
Released in [v1.5.0](https://github.com/eclipse-cdt-cloud/vscode-peripheral-inspector/actions/runs/7152420161)
